standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/properties/SerializationProxy.java

```diff
@@ -0,0 +1,614 @@
+  protected Type<T> readType(DataInput in) throws IOException {
+    final Type<T> type;
+    String className = "?";
+    int slot = (int) in.readByte() & 0xff;
+    if (slot == 255) {
+      className = in.readUTF();
+      type = (Type<T>) TYPES.computeIfAbsent(className, this::createType);
+    } else if (slot < REGISTERED.size()) {
+      type = (Type<T>) REGISTERED.get(slot);
+    } else {
+      type = null;
+    }
+    if (type == null) {
+      throw new ProxyException("can not resolve class @ " + slot +", " + className);
+    }
+    return type;
+  }
+
+  /**
+   * Encapsulates the mandatory constructor and write methods for a given proxified class.
+   * @param <T> the proxified class
+   */
+  protected static class Type<T extends Serializable> {
+    private final Constructor<T>[] ctors;
+    private final Method[] writes;
+    private transient int slot = 255;
+
+    /**
+     * Creates a new instance of type.
+     * @param clazz the proxified class
+     */
+    public Type(Class<T> clazz) {
+        ctors = typeConstructors(clazz);
+        writes = typeWrites(clazz);
+    }
+
+    /**
+     * The slot number if the class is registered.
+     * @return the slot number, 255 means not-registered
+     */
+    public int getSlot() {
+      return slot;
+    }
+
+    /**
+     * @return the target class
+     */
+    public String getTargetName() {
+      // there is always at least one ctor
+      return ctors[0].getDeclaringClass().getName();
+    }
+
+    /**
+     * Compare parameter signatures of executables.
+     * @param lhs left-hand side
+     * @param rhs right-hand side
+     * @return 0 if equal, +/- 1 if left &lt;/&gt; than right
+     */
+    private static int compareSignatures(Executable lhs, Executable rhs) {
+      return compareSignatures(lhs.getParameterTypes(), rhs.getParameterTypes());
+    }
+
+    /**
+     * Compare executables parameter signatures.
+     * @param lhs left-hand side executable
+     * @param rhs right-hand side executable
+     * @return 0 if equal, +/- 1 if left &lt;/&gt; than right
+     */
+    private static int compareSignatures(Class<?>[] lhs, Class<?>[] rhs) {
+      if (lhs.length < rhs.length) {
+        return -1;
+      }
+      if (lhs.length > rhs.length) {
+        return 1;
+      }
+      int cmp = 0;
+      // lhs.length == rhs.length
+      final int length = lhs.length;
+      for (int p = 0; p < length; ++p) {
+        Class<?> actual = lhs[p];
+        Class<?> formal = rhs[p];
+        if (formal != null && actual != null && !formal.isAssignableFrom(actual)) {
+          // if formal parameter is primitive and actual argument is compatible
+          int dist;
+          if (formal.isPrimitive() && (dist = CONVERTIBLES.get(formal).indexOf(actual)) >= 0) {
+            cmp +=  dist;
+            continue;
+          }
+          dist = formal.getName().compareTo(actual.getName());
+          if (dist != 0) {
+            return cmp * (length - p);
```

Comment:
nit: If there is a case `cmp * (length - p)` = 0 though the args mismatch?

Reply:
Should be `dist *  (length - p)`, good catch.